### PR TITLE
[Impeller] Cache location in metadata.

### PR DESCRIPTION
**Pull request by jonahwilliams**
Hacky mutable stuff, but avoids recomputing string keys and hashing them multiple times per draw call. [See also
](https://docs.google.com/document/d/1msDN8D4_Hi2zR5JFaFLWDb0IHD3MekuuiaYLJ3Tt6to/edit?resourcekey=0-bZbENKIpXKtrnzU6kkLXcQ)

While we cache the uniform location, looking up this value requires computing a string key for every single uniform value on every single command. We can speed this up by stashing the uniform location in the shader metadata, which shouldn't change for the app lifetime. This avoids both the string computation and the hashmap usage.

### CODE REVIEW

File: impeller/core/shader_types.h [hunk 1]
```diff
@@ -71,11 +71,23 @@ struct ShaderStructMemberMetadata {
   size_t size;
   size_t byte_length;
   std::optional<size_t> array_elements;
+  /// @brief This field is used to cache the runtime location of this
+  ///        Uniform data after the first time the shader metadata is used in
+  ///        the OpenGLES backend.
+  ///
+  ///        See buffer_binding_gles.cc.
+  mutable std::optional<int> location = std::nullopt;
```
[the remark this answers is not in the document]
Comment (chinmaygarde) [reply impeller/core/shader_types.h:79]: From Triage: lets just store it in the "PSO" instead.

**Comment by jonahwilliams** [comment]
I've updated this to cache the binding data in in the buffer_binding, I couldn't put them in the pipeline itself as that doesn't actually have all of the information needed. To support this I made the buffer_binding pointer non-const.

So we still use one hashmap lookup per shader, but the key we use is already generated on the shader metadata so its less work.

File: impeller/renderer/backend/gles/buffer_bindings_gles.cc [hunk 2]
```diff
@@ -194,35 +243,24 @@ bool BufferBindingsGLES::BindUniformBuffer(const ProcTableGLES& gl,
...
 
-  for (const auto& member : metadata->members) {
-    if (member.type == ShaderType::kVoid) {
-      // Void types are used for padding. We are obviously not going to find
-      // mappings for these. Keep going.
+  const auto& locations = ComputeUniformLocations(metadata);
+  for (auto i = 0u; i < metadata->members.size(); i++) {
+    const auto& member = metadata->members[i];
+    auto location = locations[i];
+    if (location == -1) {
```
Comment (chinmaygarde) [inline impeller/renderer/backend/gles/buffer_bindings_gles.cc:250]: Preserve the comment about inactive uniformed being optimized out?

Comment (chinmaygarde) [reply impeller/renderer/backend/gles/buffer_bindings_gles.cc:250]: Oh, I see it up top. Maybe just "void type or inactive".

**Review by chinmaygarde — APPROVED** [review]
I couldn't tell what information was missing to put this in the pipeline state object. But this is an improvement for sure.

**Comment by jonahwilliams** [comment]
The BufferBinding is essentially part of the PSO, so its all approximately the same, right 😄

File: impeller/renderer/backend/gles/buffer_bindings_gles.cc (hunk 2 above)
Comment (jonahwilliams) [reply impeller/renderer/backend/gles/buffer_bindings_gles.cc:251]: Done

**Pull request merged**